app/data_model/relationships_store.py

```diff
@@ -0,0 +1,92 @@
+from dataclasses import dataclass, asdict
+from typing import List, Mapping
+
+
+@dataclass
+class Relationship:
+    """
+    Represents a relationship between two items.
+    """
+
+    from_list_item_id: str
+    to_list_item_id: str
+    relationship: str
+
+    def for_json(self) -> Mapping:
+        return asdict(self)
+
+
+class RelationshipStore:
+    """
+    Stores and updates relationships.
+    """
+
+    def __init__(self, relationships: List[Mapping] = None) -> None:
+        self._is_dirty = False
+        self._relationships = (
+            [
+                Relationship(
+                    relationship['from_list_item_id'],
+                    relationship['to_list_item_id'],
+                    relationship['relationship'],
+                )
+                for relationship in relationships
+            ]
+            if relationships
+            else []
+        )
+
+    def __getitem__(self, index):
+        return self._relationships[index]
+
+    def __len__(self):
+        return len(self._relationships)
+
+    @property
+    def is_dirty(self):
+        return self._is_dirty
+
+    def clear(self):
+        self._relationships.clear()
+        self._is_dirty = True
+
+    def serialise(self):
+        return [relationship.for_json() for relationship in self._relationships]
+
+    def get_relationship(self, from_list_item_id, to_list_item_id):
+        for relationship in self._relationships:
+            if (
+                relationship.from_list_item_id == from_list_item_id
+                and relationship.to_list_item_id == to_list_item_id
+            ):
+                return relationship
+        return None
+
+    def add_or_update(self, relationship: Relationship):
+        relationship_to_update = self.get_relationship(
+            relationship.from_list_item_id, relationship.to_list_item_id
+        )
+        if relationship_to_update:
+            if relationship.relationship != relationship_to_update.relationship:
```

Comment:
Can this `if` just be part of the first `if` since it has no `else` clause:
```python3
if relationship_to_update and relationship.relationship != relationship_to_update.relationship:
```

Reply:
This has been changed to be consistent with the answer store remove.